RecoEgamma/PhotonIdentification/plugins/PhotonMVAEstimatorRun2Spring16NonTrig.cc

```diff
@@ -0,0 +1,342 @@
+         && rho.isValid() ) )
+    throw cms::Exception("MVA failure: ")
+      << "Failed to retrieve event content needed for this MVA" 
+      << std::endl
+      << "Check python MVA configuration file and make sure all needed"
+      << std::endl
+      << "producers are running upstream" << std::endl;
+
+  // Try to cast the particle into a reco particle.
+  // This should work for both reco and pat.
+  const edm::Ptr<reco::Photon> phoRecoPtr(particle);
+  if( phoRecoPtr.isNull() )
+    throw cms::Exception("MVA failure: ")
+      << " given particle is expected to be reco::Photon or pat::Photon," << std::endl
+      << " but appears to be neither" << std::endl;
+
+  // Both pat and reco particles have exactly the same accessors.
+  auto superCluster = phoRecoPtr->superCluster();
+  // Full 5x5 cluster shapes. We could take some of this directly from
+  // the photon object, but some of these are not available.
+  float e2x2 = std::numeric_limits<float>::max();
+  float e5x5 = std::numeric_limits<float>::max();
+  float full5x5_sigmaIetaIeta = std::numeric_limits<float>::max();
+  float full5x5_sigmaIetaIphi = std::numeric_limits<float>::max();
+  float effSigmaRR = std::numeric_limits<float>::max();
+
+  AllVariables allMVAVars;
+
+  const auto& full5x5_pss = phoRecoPtr->full5x5_showerShapeVariables();
+  e2x2 = full5x5_pss.e2x2;
+  e5x5 = full5x5_pss.e5x5;
+  full5x5_sigmaIetaIeta = full5x5_pss.sigmaIetaIeta;
+  full5x5_sigmaIetaIphi = full5x5_pss.sigmaIetaIphi;
+  effSigmaRR = full5x5_pss.effSigmaRR;
+
+  allMVAVars.scPhi           = superCluster->phi();
+  allMVAVars.varR9           = phoRecoPtr->r9() ;
+  allMVAVars.varSieie        = full5x5_sigmaIetaIeta; 
+  allMVAVars.varSieip        = full5x5_sigmaIetaIphi;
+  allMVAVars.varE2x2overE5x5 = e2x2/e5x5;
+  allMVAVars.varSCEta        = superCluster->eta(); 
+  allMVAVars.varRawE         = superCluster->rawEnergy(); 
+  allMVAVars.varSCEtaWidth   = superCluster->etaWidth(); 
+  allMVAVars.varSCPhiWidth   = superCluster->phiWidth(); 
+  allMVAVars.varESEnOverRawE = superCluster->preshowerEnergy() / superCluster->rawEnergy();
+  allMVAVars.varESEffSigmaRR = effSigmaRR;
+  allMVAVars.varRho          = *rho; 
+  allMVAVars.varPhoIsoRaw    = (*phoPhotonIsolationMap)[phoRecoPtr];  
+  allMVAVars.varChIsoRaw     = (*phoChargedIsolationMap)[phoRecoPtr];
+  allMVAVars.varWorstChRaw   = (*phoWorstChargedIsolationMap)[phoRecoPtr];
+
+  //photon iso corrected:
+
+  double eA = effectiveAreas_.getEffectiveArea( abs(superCluster->eta()) );
+  double phoIsoPtScalingCoeffVal = 0;
+  if( !isEndcapCategory( findCategory ( particle ) ) ) 
+    phoIsoPtScalingCoeffVal = phoIsoPtScalingCoeff_.at(0); // barrel case
+  else
+    phoIsoPtScalingCoeffVal =  phoIsoPtScalingCoeff_.at(1); //endcap case
+  double phoIsoCorr = (*phoPhotonIsolationMap)[phoRecoPtr] - eA*(*rho) - phoIsoPtScalingCoeffVal*phoRecoPtr->pt();
+
+  allMVAVars.varPhoIsoCorr = TMath::Max(phoIsoCorr, phoIsoCutoff_);
```

Comment:
please use std::max instead (general comment against use of TMath:: if a native c++ method is available)